codegen/smithy-python-codegen/src/main/java/software/amazon/smithy/python/codegen/StructureGenerator.java

```diff
@@ -121,120 +134,83 @@ private void renderError() {
-                        writer.write(formatString, memberName, memberSymbol, getDefaultValue(writer, member));
-                    }
-                }
+            var requiresField = false;
+            if (member.hasTrait(SensitiveTrait.class)) {
+                writer.putContext("sensitive", true);
+                writer.addStdlibImport("dataclasses", "field");
+                requiresField = true;
+            } else {
+                writer.putContext("sensitive", false);
             }
-        });
-
-        writer.indent();
 
-        writeClassDocs(isError);
-        if (isError) {
-            writer.write("super().__init__(message)");
-        }
+            var defaultValue = "None";
+            var defaultKey = "default";
+            if (member.hasTrait(DefaultTrait.class)) {
+                defaultValue = getDefaultValue(writer, member);
+                if (Set.of("list", "dict").contains(defaultValue)) {
+                    writer.addStdlibImport("dataclasses", "field");
+                    defaultKey = "default_factory";
+                    requiresField = true;
+                }
 
-        Stream.concat(requiredMembers.stream(), optionalMembers.stream()).forEach(member -> {
-            String memberName = symbolProvider.toMemberName(member);
-            if (isOptionalDefault(member)) {
-                writer.write("self.$1L = $1L if $1L is not None else $2L",
-                    memberName, getDefaultValue(writer, member));
             } else {
-                writer.write("self.$1L = $1L", memberName);
+                writer.putContext("nullable", true);
             }
-        });
-        writer.dedent();
-        writer.write("");
-    }
+            writer.putContext("defaultKey", defaultKey);
+            writer.putContext("defaultValue", defaultValue);
+            writer.putContext("useField", requiresField);
 
-    private boolean isOptionalDefault(MemberShape member) {
-        // If a member with a default value isn't required, it's optional.
-        // see: https://smithy.io/2.0/spec/type-refinement-traits.html#smithy-api-default-trait
-        var target = model.expectShape(member.getTarget());
-        return member.hasTrait(DefaultTrait.class) && !member.hasTrait(RequiredTrait.class)
-            && (target.isDocumentShape() || target.isListShape() || target.isMapShape());
+            var target = model.expectShape(member.getTarget());
+            writer.putContext("quote", recursiveShapes.contains(target) ? "'" : "");
+
+            var memberName = symbolProvider.toMemberName(member);
+            writer.write("""
+                    $L: ${quote:L}$T${?nullable} | None${/nullable}${quote:L} \
+                    = ${^useField}${defaultValue:L}${/useField}\
+                    ${?useField}\
+                    field(${?sensitive}repr=False, ${/sensitive}${defaultKey:L}=${defaultValue:L})\
+                    ${/useField}""", memberName, symbolProvider.toSymbol(member));
```

Comment:
This looks pretty intense, do you have an example of the generated code? Ideally we're no longer using `\` line continuations if possible.

Reply:
You can't really get away with not using the continuations without making it even harder to read. This will produce things like:

```python
@dataclass(kw_only=True)
class Foo:
    normal: int
    optional: int | None = None

    recursive: "Foo"
    optional_recursive: "Foo | None" = None

    default: int = 0
    default_list: list[str] = field(default_factory=list)
    default_map: dict[str, str] = field(default_factory=map)

    sensitive: int = field(repr=False)
    sensitive_default: int = field(default=0, repr=False)
    sensitive_default_collection: list[str] = field(default_factory=list, repr=False)
```

It's messy because all these things can combine.